### PR TITLE
[move-2024] Method syntax for `sui-framework`

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

Stacked on #16466  (and also #16634 as a result)

This translates `sui-framework` to Move 2024.

Public use funs:

```
address.move
11:    public use fun sui::object::id_from_address as address.to_id;

object.move
13:    public use fun id_to_address as ID.to_address;
16:    public use fun id_to_bytes as ID.to_bytes;
19:    public use fun uid_as_inner as UID.as_inner;
22:    public use fun uid_to_inner as UID.to_inner;
25:    public use fun uid_to_address as UID.to_address;
28:    public use fun uid_to_bytes as UID.to_bytes;

kiosk/kiosk.move
100:    public use fun kiosk_owner_cap_for as KioskOwnerCap.`for`;
```

Should we consider `public use fun` some of the definitions in `pay` for `coin`?

## Test Plan 

Tests are unchanged. Note that many of the tests may be rewritten to also use method syntax, but this change is already large enough that it seemed prudent to do the tests in a separate diff.

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/packages/sui-framework/sources/hex.move [hunk 1]
```diff
@@ -34,12 +30,11 @@ module sui::hex {
...
     /// Aborts if the hex string contains non-valid hex characters (valid characters are 0 - 9, a - f, A - F)
     public fun decode(hex: vector<u8>): vector<u8> {
-        let (mut i, mut r, l) = (0, vector[], vector::length(&hex));
+        let (mut i, mut r, l) = (0, vector[], hex.length());
         assert!(l % 2 == 0, EInvalidHexLength);
         while (i < l) {
-            let decimal = (decode_byte(*vector::borrow(&hex, i)) * 16) +
-                          decode_byte(*vector::borrow(&hex, i + 1));
-            vector::push_back(&mut r, decimal);
+            let decimal = decode_byte(hex[i]) * 16 + decode_byte(hex[i + 1]);
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/hex.move:36]: 🎉

File: crates/sui-framework/packages/sui-framework/sources/crypto/groth16.move [hunk 2]
```diff
@@ -48,11 +47,11 @@ module sui::groth16 {
 
     /// Returns bytes of the four components of the `PreparedVerifyingKey`.
     public fun pvk_to_bytes(pvk: PreparedVerifyingKey): vector<vector<u8>> {
-        let mut res = vector::empty();
-        vector::push_back(&mut res, pvk.vk_gamma_abc_g1_bytes);
-        vector::push_back(&mut res, pvk.alpha_g1_beta_g2_bytes);
-        vector::push_back(&mut res, pvk.gamma_g2_neg_pc_bytes);
-        vector::push_back(&mut res, pvk.delta_g2_neg_pc_bytes);
+        let mut res = vector[];
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/crypto/groth16.move:50]: 🙏

File: crates/sui-framework/packages/sui-framework/sources/crypto/groth16.move [hunk 3]
```diff
@@ -48,11 +47,11 @@ module sui::groth16 {
...
+        res.push_back(pvk.vk_gamma_abc_g1_bytes);
+        res.push_back(pvk.alpha_g1_beta_g2_bytes);
+        res.push_back(pvk.gamma_g2_neg_pc_bytes);
+        res.push_back(pvk.delta_g2_neg_pc_bytes);
         res
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/crypto/groth16.move:55]: Actually
```suggestion
        vector[
            pvk.vk_gamma_abc_g1_bytes,
            pvk.alpha_g1_beta_g2_bytes,
            pvk.gamma_g2_neg_pc_bytes,
            pvk.delta_g2_neg_pc_bytes,
        ]
```

File: crates/sui-framework/packages/sui-framework/sources/address.move [hunk 4]
```diff
@@ -1,12 +1,15 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+#[defines_primitive(address)]
 module sui::address {
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/address.move:5]: Can we make `std::address` the definer instead? this feels a bit strange.

Or do you want to make that a separate PR?

File: crates/sui-framework/packages/sui-framework/sources/transfer.move [hunk 5]
```diff
@@ -4,7 +4,7 @@
 #[allow(unused_const)]
 module sui::transfer {
 
-    use sui::object::{Self, ID, UID};
+    use sui::object::{ID, UID};
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/transfer.move:7]:
```suggestion
```

File: crates/sui-framework/packages/sui-framework/sources/address.move (hunk 4 above)
Comment (cgswords) [reply crates/sui-framework/packages/sui-framework/sources/address.move:5]: I would love to, but I talked with Tim some and it seems like migrating the native functions is going to be a large and arduous ordeal (plus we'll have to retain them in some way in both places to avoid breaking existing code). I thought it would be best to do it as a follow-up diff (though it's possible marking this as the primitive defining and providing the use fun here will lead to more dragons).

File: crates/sui-framework/packages/sui-framework/sources/authenticator_state.move [hunk 6]
```diff
@@ -310,19 +310,19 @@ module sui::authenticator_state {
...
             let cur_iss = &cur.jwk_id.iss;
             if (option::is_none(&prev_issuer)) {
                 option::fill(&mut prev_issuer, *cur_iss);
-                vector::push_back(&mut issuer_max_epochs, cur.epoch);
+                issuer_max_epochs.push_back(cur.epoch);
             } else {
-                if (cur_iss == option::borrow(&prev_issuer)) {
-                    let back = vector::length(&issuer_max_epochs) - 1;
-                    let prev_max_epoch = vector::borrow_mut(&mut issuer_max_epochs, back);
+                if (cur_iss == prev_issuer.borrow()) {
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/authenticator_state.move:319]: Did we land on no index syntax for option?

File: crates/sui-framework/packages/sui-framework/sources/balance.move [hunk 7]
```diff
@@ -5,7 +5,7 @@
 /// module to allow balance operations and can be used to implement
 /// custom coins with `Supply` and `Balance`s.
 module sui::balance {
-    use sui::tx_context::{Self, TxContext};
+    use sui::tx_context::TxContext;
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/balance.move:8]:
```suggestion
```

File: crates/sui-framework/packages/sui-framework/sources/bcs.move [hunk 8]
```diff
@@ -22,19 +22,18 @@
...
 ///
 ///     // unpack bcs struct
-///     let leftovers = bcs::into_remainder_bytes(prepared);
+///     let leftovers = prepared.into_remainder_bytes();
 ///
 ///     (u8_value, u64_value, leftovers)
 /// }
 /// ```
 module sui::bcs {
     use std::option::{Self, Option};
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/bcs.move:36]:
```suggestion
```

File: crates/sui-framework/packages/sui-framework/sources/clock.move [hunk 9]
```diff
@@ -6,7 +6,7 @@
 module sui::clock {
     use sui::object::{Self, UID};
     use sui::transfer;
-    use sui::tx_context::{Self, TxContext};
+    use sui::tx_context::TxContext;
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/clock.move:9]:
```suggestion
```

File: crates/sui-framework/packages/sui-framework/sources/clock.move [hunk 10]
```diff
@@ -6,7 +6,7 @@
 module sui::clock {
     use sui::object::{Self, UID};
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/clock.move:7]:
```suggestion
```

File: crates/sui-framework/packages/sui-framework/sources/package.move [hunk 11]
```diff
@@ -239,7 +236,7 @@ module sui::package {
         policy: u8,
         digest: vector<u8>
     ): UpgradeTicket {
-        let id_zero = object::id_from_address(@0x0);
+        let id_zero = @0x0.to_id();
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/package.move:239]: 🔥

File: crates/sui-framework/packages/sui-framework/sources/table_vec.move [hunk 12]
```diff
@@ -24,82 +24,108 @@ module sui::table_vec {
...
-        pop_back(t)
+        assert!(t.length() > i, EIndexOutOfBound);
+        let last_idx = t.length() - 1;
+        t.swap(i, last_idx);
+        t.pop_back()
     }
+
+    #[test]
+    fun test_swap() {
+        let ctx = &mut sui::tx_context::dummy();
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/table_vec.move:110]: Do we not have a separate `test` file for this?

File: crates/sui-framework/packages/sui-framework/sources/token.move [hunk 13]
```diff
@@ -621,48 +608,60 @@ module sui::token {
...
+    public fun spend_action(): String {
+        let spend_str = SPEND;
+        spend_str.to_string()
+    }
 
     /// Name of the `ToCoin` action.
-    public fun to_coin_action(): String { string::utf8(TO_COIN) }
+    public fun to_coin_action(): String {
+        let to_coin_str = TO_COIN;
+        to_coin_str.to_string()
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/token.move:657]: Does this not work with `TO_COIN.to_string()`? Or do we hit the warning for copying a constant? I feel like we shouldn't since its not borrowing it?

**Review by tnowacki — COMMENTED** [review]
I started leaving some comments about unnecessary `use`... but we can do that in another PR. I'll stop

**Comment by damirka** [comment]
Love this! 

> Should we consider public use fun some of the definitions in pay for coin?

I think we should go for it! Pay is one of those modules which some people will never find but would want its features. So adding them to Coin is great IMO

@cgswords @todd-mystenlabs 

Should we add the `ctx.fresh_uid()` as a part of this PR or do you think it should be done separately?

File: crates/sui-framework/packages/sui-framework/sources/table_vec.move (hunk 12 above)
Comment (cgswords) [reply crates/sui-framework/packages/sui-framework/sources/table_vec.move:110]: Evidently not?

File: crates/sui-framework/packages/sui-framework/sources/token.move (hunk 13 above)
Comment (cgswords) [reply crates/sui-framework/packages/sui-framework/sources/token.move:657]: We hit the copy issue, yeah. I'm not too worried because it does make it clear the constant is being copied. Unfortunately, it actually is borrowing it by lowering it to `TempBorrow` with a `Dereference.` I am happy to revisit the semantics of path resolution if you'd like, since this is the second time this has come up.

File: crates/sui-framework/packages/sui-framework/sources/random.move [hunk 14]
```diff
@@ -129,6 +124,6 @@ module sui::random {
         new_bytes: vector<u8>,
         ctx: &TxContext,
     ) {
-        update_randomness_state(self, new_round, new_bytes, ctx);
+        self.update_randomness_state(new_round, new_bytes, ctx);
```
Comment (tzakian) [inline crates/sui-framework/packages/sui-framework/sources/random.move:127]: Not sure what happened below but there seems to be code missing?

File: crates/sui-framework/packages/sui-framework/sources/crypto/poseidon.move [hunk 15]
```diff
@@ -23,16 +22,16 @@ module sui::poseidon {
...
     public fun poseidon_bn254(data: &vector<u256>): u256 {
-        let (mut i, mut b, l) = (0, vector[], vector::length(data));
+        let (mut i, mut b, l) = (0, vector[], data.length());
         assert!(l > 0, EEmptyInput);
         while (i < l) {
-            let field_element = vector::borrow(data, i);
+            let field_element = &data[i];
             assert!(*field_element < BN254_MAX, ENonCanonicalInput);
-            vector::push_back(&mut b, bcs::to_bytes(vector::borrow(data, i)));
+            b.push_back(bcs::to_bytes(&data[i]));
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/crypto/poseidon.move:30]: Not possible currently (I think)

File: crates/sui-framework/packages/sui-framework/sources/deny_list.move [hunk 16]
```diff
@@ -52,7 +52,8 @@ module sui::deny_list {
         `type`: vector<u8>,
         addr: address,
     ) {
-        per_type_list_add(bag::borrow_mut(&mut deny_list.lists, per_type_index), `type`, addr)
+        let bag_entry: &mut PerTypeList = &mut deny_list.lists[per_type_index];
+        bag_entry.per_type_list_add(`type`, addr)
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/deny_list.move:52]: I think it should be?

File: crates/sui-framework/packages/sui-framework/sources/display.move [hunk 17]
```diff
@@ -177,7 +176,7 @@ module sui::display {
         let uid = object::new(ctx);
 
         event::emit(DisplayCreated<T> {
-            id: object::uid_to_inner(&uid)
+            id: uid.to_inner()
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/display.move:176]: I'd vote to_inner, just to keep it consistent with the existing name

File: crates/sui-framework/packages/sui-framework/sources/hex.move [hunk 18]
```diff
@@ -15,13 +14,10 @@ module sui::hex {
...
     public fun encode(bytes: vector<u8>): vector<u8> {
-        let (mut i, mut r, l) = (0, vector[], vector::length(&bytes));
+        let (mut i, mut r, l) = (0, vector[], bytes.length());
         let hex_vector = HEX;
         while (i < l) {
-            vector::append(
-                &mut r,
-                *vector::borrow(&hex_vector, (*vector::borrow(&bytes, i) as u64))
-            );
+            r.append(hex_vector[(bytes[i] as u64)]);
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/hex.move:20]: Yeah, I haven't put up that PR

File: crates/sui-framework/packages/sui-framework/sources/kiosk/transfer_policy.move [hunk 19]
```diff
@@ -139,7 +138,7 @@ module sui::transfer_policy {
     entry fun default<T>(pub: &Publisher, ctx: &mut TxContext) {
         let (policy, cap) = new<T>(pub, ctx);
         sui::transfer::share_object(policy);
-        sui::transfer::transfer(cap, sender(ctx));
+        sui::transfer::transfer(cap, ctx.sender());
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/transfer_policy.move:138]: Not possible (and maybe not desirable? unsure)

We could relax this restriction if we wanted, but I didn't want to get into a situation where you do something like 

```
public struct NotAnObject {}
public use fun transfer::transfer as NotAnObject.transfer;
```
We lack the ability to reject this `use fun` at it's declaration, and as such, you will get an error at each usage. Which felt bad, especially for `public use fun`

File: crates/sui-framework/packages/sui-framework/sources/random.move (hunk 14 above)
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/random.move:127]: Where?

Comment (tzakian) [reply crates/sui-framework/packages/sui-framework/sources/random.move:127]: OK, this is really weird -- in my PR review (using a github client inside vim) it showed basically the whole rest of this file deleted...but that's obviously not the case now is it...

**Review by tnowacki — APPROVED** [review]

File: crates/sui-framework/packages/sui-framework/sources/deny_list.move (hunk 16 above)
[the remark this answers is not in the document]
Comment (cgswords) [reply crates/sui-framework/packages/sui-framework/sources/deny_list.move:52]: It can't figure out the type of the subject for the method call if we do it the way Tim suggested.

File: crates/sui-framework/packages/sui-framework/sources/display.move (hunk 17 above)
[the remark this answers is not in the document]
Comment (cgswords) [reply crates/sui-framework/packages/sui-framework/sources/display.move:176]: We can also come back and add those later. I think `to_id` does make sense, but this is "enough" for now.
